[system][createStyled] Intercept `ownerState` coming from `props` and `ownerState`

DiegoAndai opened the pull request:
From https://github.com/mui/material-ui/pull/42345#discussion_r1611224511

This resolves the styled symptoms of https://github.com/mui/material-ui/issues/42184 by not allowing `ownerState` values coming from `props` or `ownerState` to override the component's `ownerState`.

It does not close https://github.com/mui/material-ui/issues/42184, as nested `ownerState` values are still being propagated.

DiegoAndai commented:
See Argos changes for the fixes

mnajdova reviewed (commented):
There are few more lines where we handle this, e.g. https://github.com/mui/material-ui/blob/65622f25ed3ce44560babb9a24f9dd7cb4e3964f/packages/mui-system/src/createStyled/createStyled.js#L51 and https://github.com/mui/material-ui/blob/65622f25ed3ce44560babb9a24f9dd7cb4e3964f/packages/mui-system/src/createStyled/createStyled.js#L55

Can we create a const at the top of the function and used that in all places?

DiegoAndai commented:
@mnajdova I refactored by creating a const and added a test.

For the occurrences you point out, there's no risk of overriding. The function signature is 

```js
function processStyleArg(callableStyle, { ownerState, ...props })
```

So `props` would never have an `ownerState` key. Because of this lines 51 and 55 won't show this issue, and we can leave them as is.

mnajdova reviewed (approved):
Test, logic and argos look good to me 👌 We should port this logic to Pigment CSS too.
